# Movie Recommendation System
In this project I will use pearson correlation method to create a movie recommendation system with collaborative filtering.

Import the required libraries

In [21]:
import pandas as pd

In [3]:
ratings = pd.read_csv('userRatings.csv')
ratings.head()

'burbs, The (1989)  (500) Days of Summer (2009)  \
0                 0.0                          0.0   
1                 0.0                          0.0   
2                 0.0                          0.0   
3                 0.0                          0.0   
4                 0.0                          0.0   

   10 Cloverfield Lane (2016)  10 Things I Hate About You (1999)  \
0                         0.0                                0.0   
1                         0.0                                0.0   
2                         0.0                                0.0   
3                         0.0                                0.0   
4                         0.0                                0.0   

   10,000 BC (2008)  101 Dalmatians (1996)  \
0               0.0                    0.0   
1               0.0                    0.0   
2               0.0                    0.0   
3               0.0                    0.0   
4               0.0                    0.0   

   101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
0                                                0.0        
1                                                0.0        
2                                                0.0        
3                                                0.0        
4                                                0.0        

   12 Angry Men (1957)  12 Years a Slave (2013)  127 Hours (2010)  ...  \
0                  0.0                      0.0               0.0  ...   
1                  0.0                      0.0               0.0  ...   
2                  0.0                      0.0               0.0  ...   
3                  5.0                      0.0               0.0  ...   
4                  0.0                      0.0               0.0  ...   

   Zack and Miri Make a Porno (2008)  Zero Dark Thirty (2012)  \
0                                0.0                      0.0   
1                                0.0                      0.0   
2                                0.0                      0.0   
3                                0.0                      0.0   
4                                0.0                      0.0   

   Zero Effect (1998)  Zodiac (2007)  Zombieland (2009)  Zoolander (2001)  \
0                 0.0            0.0                0.0               0.0   
1                 0.0            0.0                3.0               0.0   
2                 0.0            0.0                0.0               0.0   
3                 0.0            0.0                0.0               0.0   
4                 0.0            0.0                0.0               0.0   

   Zootopia (2016)  eXistenZ (1999)  xXx (2002)  ¡Three Amigos! (1986)  
0              0.0              0.0         0.0                    4.0  
1              0.0              0.0         0.0                    0.0  
2              0.0              0.0         0.0                    0.0  
3              0.0              0.0         0.0                    0.0  
4              0.0              0.0         0.0                    0.0  

[5 rows x 2269 columns]

We can see that there are 2269 columns meaning there are 2269 movies in the dataset.

In [4]:
ratings.sample(5)

'burbs, The (1989)  (500) Days of Summer (2009)  \
75                  0.0                          0.0   
144                 0.0                          0.0   
178                 0.0                          0.0   
367                 0.0                          0.0   
342                 0.0                          0.0   

     10 Cloverfield Lane (2016)  10 Things I Hate About You (1999)  \
75                          0.0                                0.0   
144                         0.0                                0.0   
178                         0.0                                0.0   
367                         0.0                                0.0   
342                         0.0                                0.0   

     10,000 BC (2008)  101 Dalmatians (1996)  \
75                0.0                    1.0   
144               0.0                    0.0   
178               0.0                    0.0   
367               0.0                    0.0   
342               0.0                    0.0   

     101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
75                                                 0.0        
144                                                0.0        
178                                                0.0        
367                                                0.0        
342                                                0.0        

     12 Angry Men (1957)  12 Years a Slave (2013)  127 Hours (2010)  ...  \
75                   0.0                      0.0               0.0  ...   
144                  0.0                      0.0               0.0  ...   
178                  0.0                      0.0               0.0  ...   
367                  0.0                      0.0               0.0  ...   
342                  4.5                      0.0               0.0  ...   

     Zack and Miri Make a Porno (2008)  Zero Dark Thirty (2012)  \
75                                 0.0                      0.0   
144                                0.0                      0.0   
178                                0.0                      0.0   
367                                0.0                      0.0   
342                                0.0                      0.0   

     Zero Effect (1998)  Zodiac (2007)  Zombieland (2009)  Zoolander (2001)  \
75                  0.0            0.0                0.0               0.0   
144                 0.0            0.0                0.0               0.0   
178                 0.0            0.0                0.0               0.0   
367                 0.0            0.0                0.0               0.0   
342                 0.0            0.0                0.0               0.0   

     Zootopia (2016)  eXistenZ (1999)  xXx (2002)  ¡Three Amigos! (1986)  
75               0.0              0.0         0.0                    0.0  
144              0.0              0.0         0.0                    0.0  
178              0.0              0.0         0.0                    0.0  
367              0.0              0.0         0.0                    0.0  
342              0.0              0.0         0.0                    0.0  

[5 rows x 2269 columns]

In [6]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Columns: 2269 entries, 'burbs, The (1989) to ¡Three Amigos! (1986)
dtypes: float64(2269)
memory usage: 10.6 MB


There are 610 ratings for 2269 movies in this dataset.

In [8]:
ratings.describe().T

count      mean       std  min  25%  50%  \
'burbs, The (1989)                 610.0  0.088525  0.560348  0.0  0.0  0.0   
(500) Days of Summer (2009)        610.0  0.252459  0.968825  0.0  0.0  0.0   
10 Cloverfield Lane (2016)         610.0  0.084426  0.562077  0.0  0.0  0.0   
10 Things I Hate About You (1999)  610.0  0.312295  1.049480  0.0  0.0  0.0   
10,000 BC (2008)                   610.0  0.075410  0.470878  0.0  0.0  0.0   
...                                  ...       ...       ...  ...  ...  ...   
Zoolander (2001)                   610.0  0.310656  1.040936  0.0  0.0  0.0   
Zootopia (2016)                    610.0  0.204098  0.887183  0.0  0.0  0.0   
eXistenZ (1999)                    610.0  0.139344  0.736293  0.0  0.0  0.0   
xXx (2002)                         610.0  0.109016  0.570916  0.0  0.0  0.0   
¡Three Amigos! (1986)              610.0  0.133607  0.661963  0.0  0.0  0.0   

                                   75%  max  
'burbs, The (1989)                 0.0  5.0  
(500) Days of Summer (2009)        0.0  5.0  
10 Cloverfield Lane (2016)         0.0  5.0  
10 Things I Hate About You (1999)  0.0  5.0  
10,000 BC (2008)                   0.0  4.0  
...                                ...  ...  
Zoolander (2001)                   0.0  5.0  
Zootopia (2016)                    0.0  5.0  
eXistenZ (1999)                    0.0  5.0  
xXx (2002)                         0.0  4.0  
¡Three Amigos! (1986)              0.0  5.0  

[2269 rows x 8 columns]

We can see that the ratings range from 0.0 to 5.0

In [7]:
def standardize(row):
    new_row = (row-row.mean())/(row.max()-row.min())
    return new_row

ratings_std = ratings.apply(standardize)
ratings_std

'burbs, The (1989)  (500) Days of Summer (2009)  \
0             -0.017705                    -0.050492   
1             -0.017705                    -0.050492   
2             -0.017705                    -0.050492   
3             -0.017705                    -0.050492   
4             -0.017705                    -0.050492   
..                  ...                          ...   
605           -0.017705                    -0.050492   
606           -0.017705                    -0.050492   
607           -0.017705                    -0.050492   
608           -0.017705                    -0.050492   
609           -0.017705                     0.649508   

     10 Cloverfield Lane (2016)  10 Things I Hate About You (1999)  \
0                     -0.016885                          -0.062459   
1                     -0.016885                          -0.062459   
2                     -0.016885                          -0.062459   
3                     -0.016885                          -0.062459   
4                     -0.016885                          -0.062459   
..                          ...                                ...   
605                   -0.016885                          -0.062459   
606                   -0.016885                          -0.062459   
607                   -0.016885                          -0.062459   
608                   -0.016885                          -0.062459   
609                    0.783115                          -0.062459   

     10,000 BC (2008)  101 Dalmatians (1996)  \
0           -0.018852              -0.047377   
1           -0.018852              -0.047377   
2           -0.018852              -0.047377   
3           -0.018852              -0.047377   
4           -0.018852              -0.047377   
..                ...                    ...   
605         -0.018852              -0.047377   
606         -0.018852              -0.047377   
607         -0.018852              -0.047377   
608         -0.018852              -0.047377   
609         -0.018852              -0.047377   

     101 Dalmatians (One Hundred and One Dalmatians) (1961)  \
0                                            -0.049508        
1                                            -0.049508        
2                                            -0.049508        
3                                            -0.049508        
4                                            -0.049508        
..                                                 ...        
605                                          -0.049508        
606                                          -0.049508        
607                                          -0.049508        
608                                          -0.049508        
609                                          -0.049508        

     12 Angry Men (1957)  12 Years a Slave (2013)  127 Hours (2010)  ...  \
0              -0.077541                -0.021129         -0.022623  ...   
1              -0.077541                -0.021129         -0.022623  ...   
2              -0.077541                -0.021129         -0.022623  ...   
3               0.922459                -0.021129         -0.022623  ...   
4              -0.077541                -0.021129         -0.022623  ...   
..                   ...                      ...               ...  ...   
605            -0.077541                -0.021129         -0.022623  ...   
606            -0.077541                -0.021129         -0.022623  ...   
607            -0.077541                -0.021129         -0.022623  ...   
608            -0.077541                -0.021129         -0.022623  ...   
609            -0.077541                -0.021129          0.777377  ...   

     Zack and Miri Make a Porno (2008)  Zero Dark Thirty (2012)  \
0                            -0.024262                -0.018852   
1                            -0.024262                -0.018852   
2                            -0.024262     

In [9]:
corrMatrix = ratings.corr(method='pearson')

The method above calls the function .corr and we set the method to pearson for the pearson correlation method.

In [10]:
corrMatrix

'burbs, The (1989)  \
'burbs, The (1989)                           1.000000   
(500) Days of Summer (2009)                  0.063117   
10 Cloverfield Lane (2016)                  -0.023768   
10 Things I Hate About You (1999)            0.143482   
10,000 BC (2008)                             0.011998   
...                                               ...   
Zoolander (2001)                             0.049897   
Zootopia (2016)                              0.003233   
eXistenZ (1999)                              0.187953   
xXx (2002)                                   0.062174   
¡Three Amigos! (1986)                        0.353194   

                                   (500) Days of Summer (2009)  \
'burbs, The (1989)                                    0.063117   
(500) Days of Summer (2009)                           1.000000   
10 Cloverfield Lane (2016)                            0.142471   
10 Things I Hate About You (1999)                     0.273989   
10,000 BC (2008)                                      0.193960   
...                                                        ...   
Zoolander (2001)                                      0.252226   
Zootopia (2016)                                       0.216007   
eXistenZ (1999)                                       0.053614   
xXx (2002)                                            0.241092   
¡Three Amigos! (1986)                                 0.125905   

                                   10 Cloverfield Lane (2016)  \
'burbs, The (1989)                                  -0.023768   
(500) Days of Summer (2009)                          0.142471   
10 Cloverfield Lane (2016)                           1.000000   
10 Things I Hate About You (1999)                   -0.005799   
10,000 BC (2008)                                     0.112396   
...                                                       ...   
Zoolander (2001)                                     0.195054   
Zootopia (2016)                                      0.319371   
eXistenZ (1999)                                      0.177846   
xXx (2002)                                           0.096638   
¡Three Amigos! (1986)                                0.002733   

                                   10 Things I Hate About You (1999)  \
'burbs, The (1989)                                          0.143482   
(500) Days of Summer (2009)                                 0.273989   
10 Cloverfield Lane (2016)                                 -0.005799   
10 Things I Hate About You (1999)                           1.000000   
10,000 BC (2008)                                            0.244670   
...                                                              ...   
Zoolander (2001)                                            0.281934   
Zootopia (2016)                                             0.050031   
eXistenZ (1999)                                             0.121029   
xXx (2002)                                                  0.130813   
¡Three Amigos! (1986)                                       0.110612   

                                   10,000 BC (2008)  101 Dalmatians (1996)  \
'burbs, The (1989)                         0.011998               0.087931   
(500) Days of Summer (2009)                0.193960               0.148903   
10 Cloverfield Lane (2016)                 0.112396               0.006139   
10 Things I Hate About You (1999)          0.244670               0.223481   
10,000 BC (2008)                           1.000000               0.234459   
...                                             ...                    ...   
Zoolander (2001)                           0.240231               0.184324   
Zootopia (2016)                            0.094773               0.054024   
eXistenZ (1999)                            0.088045               0.047804   
xXx (2002)                                 0.203002               0.156932   
¡Three Amigos! (1986)                      0.083518  

We can now see the corresponding correlation values in the data. A higher rating between 2 movies mean they are highly correlated and are similar therefore we can recommend that movie to the user.

In [11]:
corrMatrix_df = pd.DataFrame(corrMatrix, index = ratings.columns, columns = ratings.columns)
corrMatrix_df

'burbs, The (1989)  \
'burbs, The (1989)                           1.000000   
(500) Days of Summer (2009)                  0.063117   
10 Cloverfield Lane (2016)                  -0.023768   
10 Things I Hate About You (1999)            0.143482   
10,000 BC (2008)                             0.011998   
...                                               ...   
Zoolander (2001)                             0.049897   
Zootopia (2016)                              0.003233   
eXistenZ (1999)                              0.187953   
xXx (2002)                                   0.062174   
¡Three Amigos! (1986)                        0.353194   

                                   (500) Days of Summer (2009)  \
'burbs, The (1989)                                    0.063117   
(500) Days of Summer (2009)                           1.000000   
10 Cloverfield Lane (2016)                            0.142471   
10 Things I Hate About You (1999)                     0.273989   
10,000 BC (2008)                                      0.193960   
...                                                        ...   
Zoolander (2001)                                      0.252226   
Zootopia (2016)                                       0.216007   
eXistenZ (1999)                                       0.053614   
xXx (2002)                                            0.241092   
¡Three Amigos! (1986)                                 0.125905   

                                   10 Cloverfield Lane (2016)  \
'burbs, The (1989)                                  -0.023768   
(500) Days of Summer (2009)                          0.142471   
10 Cloverfield Lane (2016)                           1.000000   
10 Things I Hate About You (1999)                   -0.005799   
10,000 BC (2008)                                     0.112396   
...                                                       ...   
Zoolander (2001)                                     0.195054   
Zootopia (2016)                                      0.319371   
eXistenZ (1999)                                      0.177846   
xXx (2002)                                           0.096638   
¡Three Amigos! (1986)                                0.002733   

                                   10 Things I Hate About You (1999)  \
'burbs, The (1989)                                          0.143482   
(500) Days of Summer (2009)                                 0.273989   
10 Cloverfield Lane (2016)                                 -0.005799   
10 Things I Hate About You (1999)                           1.000000   
10,000 BC (2008)                                            0.244670   
...                                                              ...   
Zoolander (2001)                                            0.281934   
Zootopia (2016)                                             0.050031   
eXistenZ (1999)                                             0.121029   
xXx (2002)                                                  0.130813   
¡Three Amigos! (1986)                                       0.110612   

                                   10,000 BC (2008)  101 Dalmatians (1996)  \
'burbs, The (1989)                         0.011998               0.087931   
(500) Days of Summer (2009)                0.193960               0.148903   
10 Cloverfield Lane (2016)                 0.112396               0.006139   
10 Things I Hate About You (1999)          0.244670               0.223481   
10,000 BC (2008)                           1.000000               0.234459   
...                                             ...                    ...   
Zoolander (2001)                           0.240231               0.184324   
Zootopia (2016)                            0.094773               0.054024   
eXistenZ (1999)                            0.088045               0.047804   
xXx (2002)                                 0.203002               0.156932   
¡Three Amigos! (1986)                      0.083518  

In [12]:
def get_similarity(movie_name, rating):
    similar_score = corrMatrix_df[movie_name]*(rating-2.5)
    
    similar_score = similar_score.sort_values(ascending = False)
    
    return similar_score

In [13]:
get_similarity("12 Years a Slave (2013)",5)

12 Years a Slave (2013)                      2.500000
Midnight in Paris (2011)                     1.248101
Dallas Buyers Club (2013)                    1.184276
Elite Squad (Tropa de Elite) (2007)          1.051170
American Sniper (2014)                       0.969024
                                               ...   
Grosse Pointe Blank (1997)                  -0.127162
Piano, The (1993)                           -0.129091
Butch Cassidy and the Sundance Kid (1969)   -0.131849
Clear and Present Danger (1994)             -0.132894
Mr. Holland's Opus (1995)                   -0.150164
Name: 12 Years a Slave (2013), Length: 2269, dtype: float64

In [18]:
action_lover = [("Zootopia (2016)", 1), ("Zoolander (2001)",5), ("xXx (2002)",5)]

In [19]:
similar_scores = pd.DataFrame()

for movie, rating in action_lover:
    similar_scores = similar_scores.append(get_similarity(movie,rating), ignore_index = True)
    
similar_scores.head()

Dances with Wolves (1990)  Clear and Present Danger (1994)  \
0                   0.158424                         0.134209   
1                   0.087093                        -0.115962   
2                   0.154017                        -0.091910   

   Piano, The (1993)  Four Weddings and a Funeral (1994)  Firm, The (1993)  \
0           0.110435                            0.109351          0.103054   
1          -0.051958                            0.228504         -0.097094   
2           0.030509                            0.088754          0.041502   

   Quiz Show (1994)  English Patient, The (1996)  Nine Months (1995)  \
0          0.094896                     0.094181            0.093959   
1         -0.052065                    -0.014246            0.032123   
2         -0.017892                     0.177440           -0.050787   

   Dead Man Walking (1995)  \
0                 0.090580   
1                -0.010619   
2                 0.104286   

   City of Lost Children, The (Cité des enfants perdus, La) (1995)  ...  \
0                                           0.087120                ...   
1                                           0.144628                ...   
2                                           0.162205                ...   

   The Lego Movie (2014)  Kingsman: The Secret Service (2015)  \
0              -0.727271                            -0.730602   
1               0.623765                             0.689255   
2               0.345530                             0.491721   

   Mad Max: Fury Road (2015)  Doctor Strange (2016)  \
0                  -0.760092              -0.772688   
1                   0.802849               0.222332   
2                   0.457322               0.487932   

   Guardians of the Galaxy (2014)  \
0                       -0.801003   
1                        0.562130   
2                        0.404359   

   Star Wars: Episode VII - The Force Awakens (2015)  Big Hero 6 (2014)  \
0                                          -0.808942          -0.823688   
1                                           0.665240           0.291431   
2                                           0.341204           0.324067   

   Deadpool (2016)  Inside Out (2015)  Zootopia (2016)  
0        -0.831296          -0.865477        -1.500000  
1         0.568037           0.540703         0.270368  
2         0.356561           0.365525         0.501904  

[3 rows x 2269 columns]

In [20]:
similar_scores.sum().sort_values(ascending = False)

Zoolander (2001)                     3.182863
xXx (2002)                           3.043942
Die Another Day (2002)               2.296979
School of Rock (2003)                2.105113
S.W.A.T. (2003)                      2.100894
                                       ...   
Postman, The (Postino, Il) (1994)   -0.217588
Coco (2017)                         -0.338090
Rise of the Guardians (2012)        -0.433151
Moana (2016)                        -0.631670
Zootopia (2016)                     -0.727727
Length: 2269, dtype: float64

We can see that the user gave ratings of 5 to xXx and Zoolander which are action and comedy movies and gave a low rating to Zootopia which is an animated movie. Therefore movies like Moana, Coco etc will not be recommended to the user. Movies like Die another day, school of rock, S.W.A.T can be recommended to the user because their similarity ratings are high and they're related(action and comedy movies).